### PR TITLE
Added zero gasPrice validation for Quorum compatibility mode.

### LINKED ISSUE #1551: Implement Quorum-compatible gas price rule
Quorum _currently_ requires a zero gas price. When a transaction is received with a non-zero gas value, the Quorum node will reject the transaction as invalid.

When intialised with the `isQuorum` genesis file config option set to true, Besu should enforce that transactions have a gas price of zero to ensure consistency.

### PR DESCRIPTION
**Pull request by mark-terry**
Signed-off-by: Mark Terry <mark.terry@consensys.net>

## PR description
Updates transaction validation to enforce zero gasPrice when `isQuorum` genesis config is set.

## Fixed Issue(s)

Fixes #1551 

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1406,6 +1406,7 @@ private void issueOptionWarnings() {
 
   private BesuCommand configure() throws Exception {
     checkPortClash();
+    checkQuorumCompatConfig();
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1409]: s/Compat/Compatible

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java [hunk 2]
```diff
@@ -94,6 +99,12 @@ public MainnetTransactionValidator(
       return signatureResult;
     }
 
+    if (quorumCompatibilityMode && !transaction.getGasPrice().isZero()) {
+      return ValidationResult.invalid(
+          TransactionInvalidReason.GAS_PRICE_MUST_BE_ZERO,
+          "gasPrice must be set to zero on a Quroum compatible network");
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:105]: Quorum typo

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java [hunk 3]
```diff
@@ -3789,4 +3791,29 @@ public void assertThatAllPortsSpecifiedSucceeds() {
...
     assertThat(commandErrorOutput.toString()).isEmpty();
   }
+
+  @Test
+  public void quorumInteropEnabledFailsWithoutGasPriceSet() throws IOException {
+    final Path genesisFile = createFakeGenesisFile(GENESIS_QUORUM_INTEROP_ENABLED);
+    parseCommand("--genesis-file", genesisFile.toString());
+    assertThat(commandErrorOutput.toString())
+        .contains(
+            "--min-gas-price must be set to zero if Quorum interop is enabled in the genesis config.");
```
Comment (macfarla) [inline besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:3801]: here we're calling it Quorum interop. TransactionValidator has "Quorum compatible". I don't have strong feelings but should be consistent.

**Review by macfarla — APPROVED** [review]
is it worth putting a TODO to say when quorum implements support for non-zero gas we can remove this restriction?

**Comment by macfarla** [comment]
as discussed would be good to get the msg about zero gasPrice propagated to RPC response (currently "invalid params")

**Comment by MadelineMurray** [comment]
> is it worth putting a TODO to say when quorum implements support for non-zero gas we can remove this restriction?

Do we have a policy against putting TODO's into the codebase? I'm neutral but if so, we could flag to Nicolas Maurice to ensure we capture this in that work.  Let me know if you want to go down that route instead of the TODO.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java (hunk 2 above)
Comment (MadelineMurray) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:105]: It's needs to be GoQuorum as well (I feel your pain on misspelling Quorum however - Nico has caught a bunch of those typos from me in the GoQuorum docs) -> Quorum now includes Hyperledger Besu.

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 3 above)
Comment (MadelineMurray) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:3801]: No strong feelings either.  If you need a tie break, I'd choose "GoQuorum compatible"

**Review by macfarla — APPROVED** [review]
overall LGTM - still some instances of Quorum that should be goQuorum

**Pull request merged**